### PR TITLE
Compute velocity using central difference

### LINKED ISSUE #2077: Use a "central difference" when calculating velocities in Servo
For lines like this:

https://github.com/ros-planning/moveit2/blob/06e5a0b29107432c8a42e1986d356193383347b8/moveit_ros/moveit_servo/src/servo_calcs.cpp#L783

h/t to @kylc

Reference:  https://dmpeli.math.mcmaster.ca/Matlab/Math4Q3/NumMethods/Lecture3-1.html

### PR DESCRIPTION
**Pull request by ibrahiminfinite**
### Description

This PR changes the calculations in `ServoCalcs` to use the central difference to compute the joint velocities more accurately.

Fixes #2077 

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)

### CODE REVIEW

**Comment by AndyZe** [comment]
My first impression from testing it out in simulation is that the behavior doesn't change noticeably. That's a good sign  :+1:

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:281]: Let's add a comment above this. Here's my suggestion, feel free to modify it.
```suggestion
  // last_joint_state_ caches the previous `original_joint_state_`
  sensor_msgs::msg::JointState internal_joint_state_, original_joint_state_, last_joint_state_;
```

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:372]: I appreciate the explanation here!

I think you should do the caching at the end of each iteration, though. That seems to flow better. Can you move this code to L513 instead?

```
  // Cache the state for a finite difference calculation in the next iteration
  // last_joint_state_ = original_joint_state_;
```

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:787]: Let's use `q(t)` and `q(t + dt)` everywhere, instead of `x(t)` and `x(t + dt)`? Because we're in the joint domain.

```suggestion
  // (q(t + dt) - q(t - dt)) / ( 2 * dt )
```

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:796]: This looks correct to me  :+1:

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:794]:
```suggestion
  // joint_state contains the future position q(t + dt)
  // last_joint_state contains past position q(t - dt)
```

**Review by AndyZe — COMMENTED** [review]
Thanks a lot! I'll see if I can get somebody to test it on a real robot.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:372]:
```suggestion
  // original_joint_state_ contains state q(t - dt)
  // internal_joint_state_ will be updated with the state q(t + dt) in this iteration.
  // last_joint_state_ will preserve the state q(t - dt) for this iteration to be used in central difference.
  // original_joint_state_ will get updated with current state x(t) in updateJoints()
```

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h
Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:281]: Thanks . I forgot to add this.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:787]: Makes sense , lets go with `q`

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:796]: Sophistication !

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (sea-bass) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: nit, but there's no need to pass basic types like double by reference here

Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: yep, it actually hurts performance (a bit...)

Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: Would it not be faster than pass by value which makes a copy ?

Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: it's a complex question but our rule of thumb is what Sebastian commented

https://blog.bytellect.com/software-development/cplusplus/which-is-faster-in-cplusplus-call-by-value-call-by-reference-or-call-by-pointer/

Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: Oh nvm, it is slower for basic types, due to unnecessary extra memory access.

Comment (sea-bass) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: tl;dr the address for basic types basically uses the same memory as the payload itself, so pass-by-reference only pays off when the data types are nontrivial

Comment (ibrahiminfinite) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:788]: Might also be good to restrict the variables captured by the lambda to just the 3 arrays it access?

Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp]: I don't think that will matter from a performance standpoint

**Review by AndyZe — APPROVED** [review]

**Comment by ibrahiminfinite** [comment]
Is this still in testing?

**Comment by AndyZe** [comment]
Waiting for a second "maintainer" to review it. It worked well when I tried it.

**Review by sjahr — APPROVED** [review]

**Pull request merged**
